homeassistant/components/vizio/media_player.py

```diff
@@ -106,127 +144,161 @@ def __init__(
         self._supported_commands = SUPPORTED_COMMANDS[device_type]
         self._device = device
         self._max_volume = float(self._device.get_max_volume())
-        self._unique_id = None
+        self._unique_id = unique_id
         self._icon = ICON[device_type]
+        self._available = True
 
     @util.Throttle(MIN_TIME_BETWEEN_SCANS, MIN_TIME_BETWEEN_FORCED_SCANS)
     async def async_update(self) -> None:
         """Retrieve latest state of the device."""
+        is_on = await self._device.get_power_state(False)
+
+        if is_on is None:
+            self._available = False
+            return
+
+        self._available = True
 
         if not self._unique_id:
-            self._unique_id = await self._device.get_esn()
+            self._unique_id = await self._device.get_esn(False)
 
-        is_on = await self._device.get_power_state()
+        if not is_on:
+            self._state = STATE_OFF
+            self._volume_level = None
+            self._current_input = None
+            self._available_inputs = None
+            return
 
-        if is_on:
-            self._state = STATE_ON
+        self._state = STATE_ON
 
-            volume = await self._device.get_current_volume()
-            if volume is not None:
-                self._volume_level = float(volume) / self._max_volume
+        volume = await self._device.get_current_volume(False)
+        if volume is not None:
+            self._volume_level = float(volume) / self._max_volume
 
-            input_ = await self._device.get_current_input()
-            if input_ is not None:
-                self._current_input = input_.meta_name
+        input_ = await self._device.get_current_input(False)
+        if input_ is not None:
+            self._current_input = input_.meta_name
 
-            inputs = await self._device.get_inputs()
-            if inputs is not None:
-                self._available_inputs = [input_.name for input_ in inputs]
+        inputs = await self._device.get_inputs(False)
+        if inputs is not None:
+            self._available_inputs = [input_.name for input_ in inputs]
 
-        else:
-            if is_on is None:
-                self._state = None
-            else:
-                self._state = STATE_OFF
+    @callback
+    @staticmethod
+    async def _async_config_entry_updated(
+        hass: HomeAssistantType, config_entry: ConfigEntry
+    ) -> None:
+        """Send update event when when Vizio config entry is updated."""
+        async_dispatcher_send(hass, UPDATE_OPTIONS_SIGNAL, config_entry)
```

Comment:
We don't need to use the dispatcher. We can update the volume here directly.

Reply:
I tried to do this directly, but I get the following error: `TypeError: 'NoneType' object is not callable`. Here is the method I am using:
```python
    async def _async_config_entry_updated(
        self, hass: HomeAssistantType, config_entry: ConfigEntry
    ) -> None:
        """Update options when when Vizio config entry is updated."""
        self._volume_step = config_entry.options[CONF_VOLUME_STEP]

    async def async_added_to_hass(self):
        """Register callbacks when entity is added."""
        # Register callback for when config entry is updated.
        self._async_unsub_listeners.append(
            self._config_entry.add_update_listener(self._async_config_entry_updated)
        )
```

Reply:
Is this because there is nothing async in the function?

Reply:
No, it's because we use weakref to create a weak reference to the callback. Then the callback can't be an instance method apparently.

Try making the callbackaa regular coroutine function but nest the callback inside `async_added_to_hass`. Then we should have access to the `self` instance anyway. I'm not too familiar with the consequences of weakref.

Reply:
same error 😞

Reply:
Ok, then let's keep what we have right now.